lib/api/model/widget.dart

```diff
@@ -0,0 +1,107 @@
+import '../../model/store.dart';
+import 'submessage.dart';
+
+/// States of a poll Zulip widget.
+///
+/// See also:
+/// - https://zulip.com/help/create-a-poll
+class Poll {
+  Poll({
+    required this.submessages,
+    required this.senderId,
+    required this.question,
+    required final List<String> options,
+  }) {
+    for (int index = 0; index < options.length; index += 1) {
+      _addOption(null, options[index], optionIndex: index);
+    }
+  }
+
+  factory Poll.fromSubmessages(
+    List<Submessage> submessages,
+    {required int senderId, required PollWidgetData widgetData}
+  ) {
+    assert(submessages.isNotEmpty);
+    final widgetEventSubmessages = submessages.sublist(1);
+
+    final poll = Poll(
+      submessages: submessages,
+      senderId: senderId,
+      question: widgetData.extraData.question,
+      options: widgetData.extraData.options,
+    );
+
+    for (final submessage in widgetEventSubmessages) {
+      final event = PollEvent.fromJson(submessage.content as Map<String, Object?>);
+      poll.applyEvent(submessage.senderId, event);
+    }
+    return poll;
+  }
+
+  /// See [Message._pollToJson].
+  final List<Submessage> submessages;
+  final int senderId;
+  String question;
+
+  Iterable<Option> get options => _options.values;
+  final Map<String, Option> _options = {};
+
+  void applyEvent(int senderId, PollEvent event) {
+    switch (event) {
+      case PollOptionEvent():
+        _addOption(
+          senderId,
+          event.option,
+          optionIndex: event.latestOptionIndex,
+        );
+
+      case PollQuestionEvent():
+        question = event.question;
+
+      case PollVoteEvent():
+        final option = _options[event.key];
+        if (option == null) return;
+
+        switch (event.op) {
+          case VoteOp.add:
+            option.voters.add(senderId);
+          case VoteOp.remove:
+            option.voters.remove(senderId);
+          case VoteOp.unknown:
+        }
+
+      case UnknownPollEvent():
+    }
+  }
+
+  void _addOption(int? senderId, String option, {required int optionIndex}) {
+    _options[PollEvent.optionKey(
+      senderId: senderId, optionIndex: optionIndex)] = Option(text: option);
+  }
+}
+
+class Option {
+  Option({required this.text});
+
+  factory Option.withVoters(String text, Iterable<int> voters) =>
+    Option(text: text)..voters.addAll(voters);
+
+  final String text;
+  final Set<int> voters = {};
+
+  Iterable<String> getVoterNames(PerAccountStore store) =>
+    voters.map((userId) => store.users[userId]?.fullName).whereType<String>();
```

Comment:
(just glancing through the changes from revision to revision, since this is still a draft)

In general when some list contains an unknown user, or an unknown stream/channel etc., we should still show that user (etc.), just with some fallback UI like the text "(unknown user)" to indicate that we don't know more details.

Otherwise we could get quite buggy-looking situations where we're showing the user misleading data — for example with a poll, we might say there were 3 votes but only list one name, or say there was just 1 vote but other users looking at the same poll see 3 votes, etc.